Add http.host parameter

aaronlehmann opened the pull request:
This allows the administrator to specify an externally-reachable URL for
the registry. It takes precedence over the X-Forwarded-Proto and
X-Forwarded-Host headers, and the hostname in the request.

NOTE: This is an alternative approach to #1000. I'm opening this as a separate PR so the patches can be compared and the best one can be chosen. Only one of the two should be merged.

stevvooe commented on registry/api/v2/urls.go:
```diff
@@ -88,6 +88,28 @@ func NewURLBuilderFromRequest(r *http.Request) *URLBuilder {
+// NewURLBuilderFromBase uses preconfigured server address information to
+// construct the root url. configScheme and configHost are the URI scheme and
+// hostname, respectively, from the configuration.
+func NewURLBuilderFromBase(r *http.Request, configScheme, configHost string) *URLBuilder {
```
I am still wildly confused why we can't just `NewURLBuidlerFromString`? That will allow one to change the full mount of the host path or will just work out of the box. For example, if http.host gets "http://foo.com", it will do the right thing and if it gets "https://foo.com/root/" it will do the thing. No need to require the request as an argument.

aaronlehmann commented:
I was a bit confused about what the path in the URL actually signified. I stared at the code a little and now it makes sense. I've updated this PR to use the full URL specified in `http.host` (so a path prefix is now allowed). I've removed `NewURLBuilderFromBase` since it's not necessary.

I'm preparsing the URL in `NewApp` to detect parse errors at startup, and avoid redundant parsing on every request. Because of this, I ended up using `NewURLBuilder` instead of `NewURLBuilderFromString`. Other than preparsing, the code is equivalent.

stevvooe commented:
LGTM

RichardScothern commented:
LGTM
